common/locks/priority_semaphore_impl.go

```diff
@@ -0,0 +1,212 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package locks
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"sync"
+)
+
+type (
+	Priority uint32
+	waiter   struct {
+		n     int
+		ready chan<- struct{} // Closed when semaphore acquired.
+	}
+	PrioritySemaphoreImpl struct {
+		size      int
+		cur       int
+		mu        sync.Mutex
+		waitLists []*list.List // Array of lists for managing different priority levels
+	}
+)
+
+const (
+	PriorityHigh Priority = iota
+	PriorityLow
+	NumPriorities
+)
+
+var _ PrioritySemaphore = (*PrioritySemaphoreImpl)(nil)
+
+// NewPrioritySemaphore creates a new semaphore with the given
+// maximum combined weight for concurrent access, capable of handling multiple priority levels.
+// Most of the logic is taken directly from golang's semaphore.Weighted.
+func NewPrioritySemaphore(n int) *PrioritySemaphoreImpl {
+	waitLists := make([]*list.List, NumPriorities)
+	for i := range waitLists {
+		waitLists[i] = list.New()
+	}
+	return &PrioritySemaphoreImpl{
+		size:      n,
+		waitLists: waitLists,
+	}
+}
+
+// Acquire acquires the semaphore with a weight of n and a priority, blocking until resources
+// are available or ctx is done. On success, returns nil. On failure, returns
+// ctx.Err() and leaves the semaphore unchanged.
+func (s *PrioritySemaphoreImpl) Acquire(ctx context.Context, priority Priority, n int) error {
+	if priority >= NumPriorities {
+		panic(fmt.Sprintf("semaphore: invalid priority %v, priority must be less than %v", priority, NumPriorities))
+	}
+
+	done := ctx.Done()
+
+	s.mu.Lock()
+	select {
+	case <-done:
+		// ctx becoming done has "happened before" acquiring the semaphore,
+		// whether it became done before the call began or while we were
+		// waiting for the mutex. We prefer to fail even if we could acquire
+		// the mutex without blocking.
+		s.mu.Unlock()
+		return ctx.Err()
+	default:
+	}
+	// Check if acquisition can proceed without waiting
+	if s.size-s.cur >= n && s.noWaiters() {
+		// Since we hold s.mu and haven't synchronized since checking done, if
+		// ctx becomes done before we return here, it becoming done must have
+		// "happened concurrently" with this call - it cannot "happen before"
+		// we return in this branch. So, we're ok to always acquire here.
+		s.cur += n
+		s.mu.Unlock()
+		return nil
+	}
+
+	if n > s.size {
+		// Don't make other Acquire calls block on one that's doomed to fail.
+		s.mu.Unlock()
+		<-ctx.Done()
+		return ctx.Err()
+	}
```

Comment:
This behavior is same as golang's Weighted semaphore. Documentation for Acquire() function says this `On failure, returns ctx.Err() and leaves the semaphore unchanged.` I guess they did this to return only ctx.Err() in any failure. I had noticed this and decided to keep this behaviour same as Weighted semaphore. Do you think we should change this for our implementation?

Reply:
They do that so too-large requests can run once the context is finished. If we're using this for bounded contexts that makes sense but I don't believe we are: this would mean a permanently blocked goroutine in the server.

I would rather we return an actual error that we can alert on (or panic if this is an unrecoverable situation) as this indicates a bug in our usage of the semaphore

Reply:
I actually think all contexts from shard context are bounded. 
But I agree this could be hard to debug if the logic just silently block here forever. We can check if there's deadline in the context and panic/return error if there's no in this case.

Reply:
Yeah it will be trick to debug. I added code to return an error if n is greater than size.